test/index.js

```diff
@@ -101,3 +124,41 @@ test('fetchConfigFromReq - source specified for request origin in header', (t) =
   t.end()
 
 })
+
+test('fetchConfigFromReq - extraHeaders specified', (t) => {
+
+  const network = 'ropsten'
+  const projectId = 'foo'
+  const extraHeaders = { 'User-Agent': 'app/1.0' }
+  const req = {
+    jsonrpc: '2.0',
+    id: 1,
+    method: 'eth_sendRawTransaction',
+    params: ['0x0102030405060708090a0b0c0d0e0f'],
+    origin: 'happydapp.eth',
+  }
+
+  const { fetchUrl, fetchParams } = fetchConfigFromReq({ network, req, projectId, extraHeaders })
+  t.equals(fetchUrl, 'https://ropsten.infura.io/v3/foo')
+  t.deepEquals(fetchParams, {
+    method: 'POST',
+    headers: {
+      'User-Agent': 'app/1.0',
+      'Accept': 'application/json',
+      'Content-Type': 'application/json',
+    },
+    body: fetchParams.body,
+  })
+  t.end()
+
+})
+
+test('createInfuraMiddleware throws with an invalid projectId', (t) => {
+
+  t.throws(() => createInfuraMiddleware({ projectId: 42 }), /Invalid value for 'projectId'/u)
+  t.throws(() => createInfuraMiddleware({ projectId: null }), /Invalid value for 'projectId'/u)
+  t.throws(() => createInfuraMiddleware({ projectId: undefined }), /Invalid value for 'projectId'/u)
+  t.throws(() => createInfuraMiddleware({ projectId: '' }), /Invalid value for 'projectId'/u)
+  t.end()
+
+})
```

Comment:
It might be worth adding a test case for invalid headers but I won't block on it.